src/passes/OptimizeInstructions.cpp

```diff
@@ -458,6 +458,94 @@ struct OptimizeInstructions
           return c;
         }
       }
+      {
+        Const* c;
+        Binary* bin;
+        Expression *x, *y;
+        BinaryOp op;
+
+        // fneg(x) * fneg(y)   ==>   x * y
+        // fneg(x) / fneg(y)   ==>   x / y
+        if (matches(
+              curr,
+              binary(&bin, Mul, unary(Neg, any(&x)), unary(Neg, any(&y)))) ||
+            matches(
+              curr,
+              binary(&bin, DivS, unary(Neg, any(&x)), unary(Neg, any(&y))))) {
+          bin->left = x;
+          bin->right = y;
+          return bin;
+        }
+        // fneg(x * C)   ==>   x * -C
+        // fneg(x / C)   ==>   x / -C
+        // fneg(C / x)   ==>   -C / x
+        //
+        // TODO: using `fastMath` is not necessary, but due to fuzzer can't
+        // flexible handle sign of NaNs we use it here.
+        if (fastMath &&
+            (matches(curr, unary(Neg, binary(&bin, Mul, any(&x), fval(&c)))) ||
+             matches(curr, unary(Neg, binary(&bin, DivS, any(&x), fval(&c)))) ||
+             matches(curr,
+                     unary(Neg, binary(&bin, DivS, fval(&c), any(&x)))))) {
+          c->value = c->value.neg();
+          return bin;
+        }
+        // (fneg(x) * y) op C   ==>   (x * y) op -C
+        // (x * fneg(y)) op C   ==>   (x * y) op -C
+        // (fneg(x) / y) op C   ==>   (x / y) op -C
+        // (x / fneg(y)) op C   ==>   (x / y) op -C
+        //
+        // where op = `*` `/` or `+`.
+        // `x` and `y` is not constants due to this may conflict
+        //  with `-x / C` or `-x * C` rules.
+        if ((matches(curr,
```

Comment:
How can these conflict? The rule here is
```
(-x * y) / C   ==>   (x * y) / -C
```
and the rule you are worried about is
```
-x / C   =>   x / -C
```
I don't see how both rules can apply to the same thing. While both have a ` / C`, the first one looks for a `*` right before the `/`, and the second rule for a `-`.

Reply:
Ok, I will try remove non-const constrains. And refuzz again

Reply:
Alright. It seems with [recent changes](https://github.com/WebAssembly/binaryen/pull/3330/commits/30885a4876cbd83bf999baa4733c92ca8303e0f2) this is not conflict anymore.